app.py

```diff
@@ -60,17 +95,70 @@ def show_in_filesystem(self, entity_type, entity_ids):
         with the given entity ids.
         """
         paths = []
-        
+
+        # As the this method relies on the path_cache to find the folders,
+        # we must check we are in sync before continuing
+        self.sgtk.synchronize_filesystem_structure()
+
         for eid in entity_ids:
+
+            if entity_type == "PublishedFile":
+                # Find the publish so that we can get the path back.
+                published_file = self.sgtk.shotgun.find_one("PublishedFile", [["id", "is", eid]], ['code', 'path'])
+                # Resolve the path for the local OS
+                try:
+                    local_path = sgtk.util.resolve_publish_path(self.sgtk, published_file)
+                except Exception as e:
+                    self.log_debug("Publish path couldn't be resolved falling back to context based path; reason: %s" % e)
+                    local_path = None
+
+                if local_path:
+                    if (os.path.isdir(local_path) or os.path.isfile(local_path)):
+                        # If the path is pointing to a valid file or folder then its good to use as is.
+                        paths.append(local_path)
+                        continue
+
+                    # possibly we have an image sequence and therefore its a symbolic path, instead check to see if
+                    # the parent folder of the path is valid and try opening that.
+                    # The issue with this logic is that possibly it was a directory that just didn't exist, as so
+                    # we would just be gathering the next directory up, ideally need a better way to handle sequences
+                    parent_dir = os.path.dirname(local_path)
+                    if parent_dir and os.path.isdir(parent_dir):
+                        paths.append(parent_dir)
+                        continue
+
             # Use the path cache to look up all paths linked to the task's entity
-            context = self.tank.context_from_entity(entity_type, eid)
-            paths.extend( context.filesystem_locations )
+            context = self.sgtk.context_from_entity(entity_type, eid)
+            context_paths = context.filesystem_locations
+
+            if context.step:
```

Comment:
Yeah, this is a good point, but one I can't really answer because I don't know what the ramifications of doing that might be.
I understand why the behaviour is not there by default, its because the step entities are non-project entities and are therefore not linked to a specific entity, so the only way I could find of finding the correct step folder, was to find all folders registered with that step and then see if it fits within the paths generated from the context. As you commented this approach could potentially be slow when dealing with lots of shots or assets (I've not tested).

Obviously, I wouldn't want to move a non-performant action into the core code. And at the moment I can't think of a better way of doing it. I added this feature (to open step folders) knowing full well that it was potentially flawed, so I'm OK with redacting this from the PR if we need to find a better solution. It would fit outside of the scope of my original task, so I would prefer to get the other fixes/features out.